third_party/BCVTB/utilXml.c

```diff
@@ -860,51 +857,72 @@ int check_variable_cfg_Validate(char const* const fileName) {
   char const* const jarPath = "/lib/xml/build/jar/DTDValidator.jar";
 #endif
 
-  BCVTB_HOME = getenv("BCVTB_HOME");
-  if (NULL == BCVTB_HOME) {
+  char const* const BCVTB_HOME = getenv("BCVTB_HOME");
+  if (BCVTB_HOME == NULL) {
     fprintf(stderr, "Error: Cannot get environment variable: BCVTB_HOME.\n");
     return -1;
   }
-  command = (char*)malloc(sizeof(char) * (strlen(BCVTB_HOME) + strlen(fileName) + strlen("variables.dtd") + 20) * 3);
-  if (NULL == command) {
-    fprintf(stderr,
-            "Error: Memory allocation failed in"
-            "       check_variable_cfg_Validate"
-            "       when parsing file '%s'.\n"
-            "       Program aborting.\n",
-            fileName);
-    return -1;
-  }
-  dtdFileName = (char*)malloc(sizeof(char) * (strlen(BCVTB_HOME) + 30));
-  if (NULL == command) {
-    fprintf(stderr,
-            "Error: Memory allocation failed in"
-            "       check_variable_cfg_Validate"
-            "       when parsing file '%s'. \n"
-            "       Program aborting.\n",
-            fileName);
+
+  size_t n_BCVTB_HOME = strlen(BCVTB_HOME);
+  size_t n_fileName = strlen(fileName);
+  size_t n_xmlPath = strlen(xmlPath);  // 9
+  size_t n_jarPath = strlen(jarPath);  // 35
+  size_t n_dtdName = strlen(dtdName);  // 13
+
+  // Verify that variables.dtd exists and is readable
+  {
+    // Add 1 for the null terminator
+    size_t n_needed = n_BCVTB_HOME + n_xmlPath + n_dtdName + 1;
+
+    char* dtdFileName = (char*)malloc(sizeof(char) * n_needed);
+    if (!dtdFileName) {
+      fprintf(stderr,
+              "Error: Memory allocation for 'dtdFileName' failed in check_variable_cfg_Validate when parsing file '%s'. \n"
+              "       Program aborting.\n",
+              fileName);
+      // free(dtdFileName);
+      return -1;
+    }
+
+    snprintf(dtdFileName, n_needed, "%s%s%s", BCVTB_HOME, xmlPath, dtdName);
+    FILE* dtdF = fopen(dtdFileName, "r");
+    if (!dtdF) {
+      fprintf(stderr, "Error: Cannot open '%s'.\n", dtdFileName);
+      free(dtdFileName);
+      return -1;
+    } else {
+      fclose(dtdF);
+    }
```

Comment:
Calculate the size needed properly, use snprintf not sprintf.

A test bed that will showcase the before/after is provided for convenience here: https://compiler-explorer.com/z/xoW58hn3j

Reply:
I'm also freeing as soon as I can, and scoping variables better.